mmv1/third_party/terraform/services/artifactregistry/data_source_artifact_registry_docker_image.go

```diff
@@ -0,0 +1,331 @@
+package artifactregistry
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-provider-google/google/tpgresource"
+	transport_tpg "github.com/hashicorp/terraform-provider-google/google/transport"
+)
+
+// https://cloud.google.com/artifact-registry/docs/reference/rest/v1/projects.locations.repositories.dockerImages#DockerImage
+type DockerImage struct {
+	name           string
+	uri            string
+	tags           []string
+	imageSizeBytes string
+	mediaType      string
+	uploadTime     string
+	buildTime      string
+	updateTime     string
+}
+
+func DataSourceArtifactRegistryDockerImage() *schema.Resource {
+
+	return &schema.Resource{
+		Read: DataSourceArtifactRegistryDockerImageRead,
+
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: `Project ID of the project.`,
+			},
+			"repository": {
```

Comment:
Did you find precedence for this sort of naming? The API itself uses a mix of terms: `repository_name`, `location` (and `registry_location`), and `docker_image` (and `image_id` and `digest`), and I mostly want to make sure that these fields will be clear to users, especially if we are parsing an API field into multiple parts.

Reply:
Hey @roaks3 I wrote a commit that removes the `project` and `location` fields and changes the `repository` field to be the fully-qualified identity of the repository.

This is inline with how the `repository` field is being used in the [google_artifact_registry_repository](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/artifact_registry_repository#repository) resource, when they use `repository` within the `upstream_policies` block.

Reply:
This is definitely reasonable, but I think I would stick with the original split, simply because after looking through a large number of GCP's datasources, this seems to be the far more common pattern. I think in particular it benefits users that have a project/region set up at the provider level. So I would suggest `project`, `location`, and `repository_id`. IIRC `location` implies a region or zone can be used, so you may want to check if a zone is valid, and then decide between `location` and `region`. `repository_id` matches the repository resource field name.

You could still make an argument for the `artifact_registry_repository_id` you're suggesting, but for consistency I think this is worth changing.

For the `docker_image` field, that name seems fine to me (`image_name` might also work), but I'm a little concerned about the `digest` field. It seems that term is more often used for the whole name + sha, so could be ambiguous in this datasource, but TBH I'm not familiar enough with how these terms are used.

Reply:
@Subserial as a member of the service team, could you weigh in on the field names used here? Do they reasonably match the API?

Reply:
Comparing this to the repository data source, `project`, `location` and `repository_id` are best for consistency.

`image`, `digest` and `tag` describe the image, but I think it would be best as mentioned below to use a single `image_name` value. The tag or digest would be appended to the image on all requests, anyways.

Reply:
@roaks3  After changing to `image_name` field from the conversation below, I've split apart the `repository` argument to be as @Subserial described: a `project`, `location`, and `repository_id`.